src/adams_methods.jl

```diff
@@ -0,0 +1,655 @@
+        for j = 1 : steporder
+            y[i+1] += dt[i]*b_imp[steporder, j]*dy[i - (steporder -1) + j]
+        end
+
+        ##(E)valulate the function anew at corrected approximatiion t[i+1],y[i+1]
+        dy[i+1] = F(t[i+1], y[i+1])
+    end
+    return t, y
+end
+
+ode_am5(F::Function,y0, t)= ode_am(F::Function,y0, t,5)
+
+
+################################################################################
+# Variable Step Variable Order (VSVO) Adams-Moulton Solver for nonstiff problems
+#
+#     [T,X] = ODE113(ODEFUN, Y0, TSPAN,ORDER=4;kwords...) with TSPAN = [T0:TFINAL]
+#     or[T0,T1,...,TFINAL] for dense output
+#
+#     (Main ref: Hairer & Wanner 1996, Vol I, p.356-360, p.398-400, p.421-423)
+################################################################################
+function ode113(F, y0, tspan, order::Integer=4; kwords...)
+    # For y0 which don't support indexing.
+    F_ = (t, y) -> [F(t, y[1])]
+    t,y = ode113(F_, [y0], tspan, order; kwords...)
+    return t, vcat_nosplat(y)
+end
+
+"a variable stepsize, variable order multi-step solver for nonstiff ODEs"
+function ode113(F, y0::AbstractVector, tspan, order::Integer=4;adaptive_order = true,
+                                                      max_order = 13,
+                                                      reltol = 1.0e-9,
+                                                      abstol = 1.0e-9,
+                                                      minstep=abs(tspan[1] - tspan[end])/1e18,
+                                                      maxstep=abs(tspan[end] - tspan[1])/2.5,
+                                                      initstep=0.,
+                                                      facmax = 1.5,
+                                                      facmin = .1,
+                                                      fac = .8,
+                                                      norm=Base.norm,
+                                                      points =:all)
+
+    #-##########################################################################
+    # Initializations
+    #-##########################################################################
+    Et, Eyf, Ty = make_consistent_types(F, y0, tspan)
+    dof = length(y0)
+
+    # initialization of time span, which will have variable length
+    tspan = convert(Vector{Et}, tspan)
+    tstart = tspan[1]
+    tfinal = tspan[end]
+    t = [tstart]
+
+    # initialzation of y[n], which will have variable length as well
+    y = Array(Ty, 1)
+    allocate!(y, y0, dof)
+    # TODO: fix after https://github.com/JuliaLang/julia/issues/16667
+    # y[1] = deepcopy(y0)
+    y[1] = copy(y0)
+
+    # initialzation of dy[n], which will have variable length as well
+    Tdy = typeof(F(t[1], y[1]))
+    dy = Tdy[]
+    push!(dy,F(t[1], y[1]))
+
+    # preallocate work variables for t_np1, y_np1 and dy_np1  which are for
+    # for t_{n+1}, y(t_{n+1}), and dy(t_{n+1})
+    t_np1   = tspan[1]
```

Comment:
For some reason I get a 

```
ERROR: LoadError: UndefVarError: t_np1 not defined
```
when I don't preallocate `t_np1`. I looked more into this, and I think it is a scope issue. Since I need `t_np1` for both all three parts of the time stepping loop (value update, step size validation, and order control) do you think it makes sense to declare the variable here?